csrc/expr_simplifier.cpp

```diff
@@ -1527,77 +1542,99 @@ bool hasCompatibleSign(Val* x, Val* y, const Context& context) {
   return isNonNegative(x, context) && isNonNegative(y, context);
 }
 
+#define CACHE_AND_RETURN(value)                                  \
+  context.less_than_cache_.emplace(std::make_pair(x, y), value); \
+  return value
+
 bool lessThan(Val* x, Val* y, const Context& context) {
+  auto cache_it = context.less_than_cache_.find({x, y});
+  if (cache_it != context.less_than_cache_.end()) {
+    return cache_it->second;
+  }
+
   x = foldConstants(x);
   y = foldConstants(y);
   if (x->value().hasValue() && y->value().hasValue()) {
-    return x->value() < y->value();
+    bool result = x->value() < y->value();
+    CACHE_AND_RETURN(result);
   }
   x = maybeUnwrapMagicZero(x);
   y = maybeUnwrapMagicZero(y);
   if (x->isZero() && isPositiveHelper(y, context)) {
-    return true;
+    CACHE_AND_RETURN(true);
   }
   // i1 % i2 < i2
   if (auto bop = dynamic_cast<BinaryOp*>(x->definition());
       bop != nullptr && bop->getBinaryOpType() == BinaryOpType::Mod) {
     auto denominator = bop->rhs();
     if (denominator->sameAs(y) && isValidDenominator(denominator, context) &&
         isNonNegative(y, context)) {
-      return true;
+      CACHE_AND_RETURN(true);
     }
   }
   // x <= a & a < b & b <= y  -->  x < y
   for (const auto& [a, b] : context.getKnownLessThan()) {
     if (lessEqual(x, a, context) && lessEqual(b, y, context)) {
-      return true;
+      CACHE_AND_RETURN(true);
     }
   }
-  return false;
+  CACHE_AND_RETURN(false);
 }
 
+#undef CACHE_AND_RETURN
+
+#define CACHE_AND_RETURN(value)                                   \
```

Comment:
It might be clearer to rename these, e.g. `CACHE_AND_RETURN_LT` and ` CACHE_AND_RETURN_LE`.

Reply:
changed